Sparse representation

bjones3 opened the pull request.

bjones3 committed:
Refactoring endpoints and GameState class
Fixes #72
Fixes #73

phblj committed:
passing tests

phblj committed:
added test coverage report

bjones3 committed:
Requested changes fixed

bjones3 committed:
Merge branch 'refactor_endpoints' of https://github.com/bjones3/webMinigame into refactor_endpoints

phblj committed:
mock out db

phblj committed:
abstracted mocks out to base test

bjones3 committed:
Added defensive copy

bjones3 committed:
Merge branch 'refactor_endpoints' of https://github.com/bjones3/webMinigame into refactor_endpoints

bjones3 committed:
Endpoint message updates

bjones3 committed:
sparse representation of game state

phblj commented on webserver.py:
```diff
@@ -41,61 +43,205 @@ def get_conn():
+        total_seed_count = 0
+        total_crops_planted = 0
+        for seed_id in GAME_CONFIG['seeds']:
+            total_seed_count += self.data['seedCounts'][seed_id]
```
This spot's going to give you trouble-- if you intialized `data['seedCounts']` with just {'a': 0}, when it counts through all the seed_ids available it's going to fail a bunch of times.

A simple fix is to use `.get()` or create a `get_seed_count(seed_id)` method on GameState

A slightly more efficient way would be to just iterate over the seeds you do have in `data['seedCounts']` because you don't care about the other seed_ids-- they're going to be 0.

```
for seed_id in self.data['seedCounts']:
    total_seed_count += self.data['seedCounts'][seed_id]
```

phblj commented on webserver.py:
```diff
@@ -41,61 +43,205 @@ def get_conn():
+                        self.data['plots'][str(i) + '_' + str(j)]['seedType'] = self.data['plot' + str(i) + '_' + str(j)]['seedType']
+                        self.data['plots'][str(i) + '_' + str(j)]['sowTime'] = self.data['plot' + str(i) + '_' + str(j)]['sowTime']
+                    del self.data['plot' + str(i) + '_' + str(j)]
+        for i in range(2):
```
long term (doesn't have to be this PR) you should pull the `2`s out of here and put it in the game config-- so the config determines how many free columns/rows you start with

phblj reviewed (approved):
I think you'll fix the only issue I found when you get tests passing

phblj commented on webserver.py:
```diff
@@ -41,61 +43,205 @@ def get_conn():
+        return message
+
+    def sell(self, seed_id):
+        if self.data['seedCounts'][seed_id] <= 0:
```
You should also be checking `if seed_id in self.data['seedCounts']` before this check.  If you make a `def get_seed_count(seed_id):` method, you won't need the extra check because `seed_id` not being in the data will yield a 0

phblj commented on webserver.py:
```diff
@@ -41,61 +43,205 @@ def get_conn():
+            if self.data['resources'][resource] < recipe_data[resource + 'Cost']:
+                message = "Not enough " + resource + " to buy a %s seed." % recipe_data['name']
+                return message
+            if self.data['seedCounts'][recipe_id] == GAME_CONFIG['max_seed_count']:
```
I'd prefer a >= check here-- if we ever end up with bad data (more seeds than max count) it'll still stop them from buying more

phblj commented on webserver.py:
```diff
@@ -41,61 +43,205 @@ def get_conn():
+                self.data['recipes'].append(recipe_id)
+            seed_id = RECIPE_CONFIG['recipes'][recipe_id]['seed_id']
+            if seed_id not in self.data['seedCounts']:
+                self.data['seedCounts'][seed_id] = 0
```
I like this initialization to 0 when the player learns the recipe!

phblj committed:
added overharvest test

bjones3 committed:
requested changes fixed

bjones3 committed:
Merge branch 'sparse_representation' of https://github.com/bjones3/webMinigame into sparse_representation

bjones3 committed:
fixed overharvest test

bjones3 committed:
Don't display unknown resource states

phblj committed:
Merge branch 'master' into sparse_representation

phblj commented on static/frontEndLayout.js:
```diff
@@ -114,35 +114,37 @@ var loadGameConfig = function() {
+        }
+        document.getElementById("cash").innerHTML = "cash: $" + state.resources.cash;
+    var numPlots = Object.keys(state.plots).length
+    buyMsg = "Click any plot to unlock - $" + GAME_CONFIG['plotPrice'] * Math.pow(GAME_CONFIG['plotMultiplier'],numPlots - 4);
```
This `4` should probably be `starting_width * starting_height`

phblj commented on static/frontEndLayout.js:
```diff
@@ -300,15 +304,15 @@ var harvest = function(x, y) {
 var unlock = function(x, y) {
-    if (state.resources.cash < GAME_CONFIG.plotPrice * Math.pow(GAME_CONFIG.plotMultiplier,state.unlockCount)) {
+    var numPlots = Object.keys(state.plots).length
+    if (state.resources.cash < GAME_CONFIG.plotPrice * Math.pow(GAME_CONFIG.plotMultiplier,numPlots - 4)) {
```
this 4 should be changed too

phblj commented on webserver.py:
```diff
@@ -218,14 +242,12 @@ def harvest(self, x, y):
 
     def unlock(self, x, y):
-        plot_price = GAME_CONFIG['plotPrice'] * GAME_CONFIG['plotMultiplier'] ** self.data['unlockCount']
+        plot_price = GAME_CONFIG['plotPrice'] * GAME_CONFIG['plotMultiplier'] ** (len(self.data['plots']) - 4)
```
4️⃣

phblj commented on static/frontEndLayout.js:
```diff
@@ -300,15 +304,15 @@ var harvest = function(x, y) {
-        document.getElementById("cash").innerHTML = "CASH: $" + state.resources.cash;
-        buyMsg = "Click any plot to unlock - $" + GAME_CONFIG['plotPrice'] * Math.pow(GAME_CONFIG['plotMultiplier'],state.unlockCount);
+        document.getElementById("cash").innerHTML = "cash: $" + state.resources.cash;
+        buyMsg = "Click any plot to unlock - $" + GAME_CONFIG['plotPrice'] * Math.pow(GAME_CONFIG['plotMultiplier'],numPlots - 4);
```
4️⃣

phblj reviewed (changes requested):
Fixes #77

phblj commented on webserver.py:
```diff
@@ -91,30 +91,37 @@ def save(self, slug):
         for resource in GAME_CONFIG['starting_resources']:
             if self.data['resources'].get(resource) is None:
-                self.data['resources'][resource] = GAME_CONFIG['starting_resources'][resource]
+                if GAME_CONFIG['starting_resources'][resource] > 0:
```
I don't think you need this "if starting resources > 0" check.  (though it doesn't hurt)

The reason is, if I set the game config to say `starting_resources: {'apple_seed': 0}` my intention may be that the user sees apple seeds initially, but just has zero of them.  with this `if` check in place, including a 0 starting resource is no different from leaving it out entirely

phblj commented on webserver.py:
```diff
@@ -91,30 +91,37 @@ def save(self, slug):
-                self.data['resources'][resource] = GAME_CONFIG['starting_resources'][resource]
+                if GAME_CONFIG['starting_resources'][resource] > 0:
+                    self.data['resources'][resource] = GAME_CONFIG['starting_resources'][resource]
+        if self.data.get('plots') is None:  # temp
```
I like marking the parts of code that are just temporary.  Making it clear where they start and end and why they're temporary will make it easier to clean up later.

phblj commented on webserver.py:
```diff
@@ -91,30 +91,37 @@ def save(self, slug):
+                    self.data['plots'][str(i) + '_' + str(j)] = {'seedType': 0, 'sowTime': 0}
         return self
 
     def check_cash(self):
```
This can be simplified a bit by getting rid of both `total_*` variables.  As soon as either of them is >0, you can `return` and abandon the method immediately.  If you get to the end without bailing out, you know you're in a situation where you need to give them a bit more cash

phblj committed:
initial plot tests

phblj reviewed (commented):
None of these comments need fixing, but I did push up a failing test.  Make sure you `git pull` before making more changes

phblj committed:
added a bunch of tests for illegal client requests

phblj committed:
left a line in on accident

bjones3 committed:
requested changes fixed
illegal client request failures fixed
cursor pointer for sow buttons/harvest button
